### PR TITLE
Chronos: add distributed api to TCNForecaster

### PR DESCRIPTION
**Pull request by TheaperDeng**
In this PR, Orca is adopted as distributed forecaster's internal implementation. 
- Users can set `distributed=True` when they init a forecaster and use the same API as local.
- Users can use `to_local()` method to seamlessly change a forecaster from distributed version (internal implemented by orca est) to a non-distributed one (internal implemented by base_pytorch_model). 
- After calling `to_local()`, users can use the distributed trained forecaster with ONNX runtime inference.

Detailed api design please refer to https://junweitestdoc.readthedocs.io/en/forecaster-dis-1/doc/PythonAPI/Chronos/forecasters.html#chronos-model-forecast-tcn-forecaster

Basic usage:
```python
# init
forecaster = TCNForecaster(..., distributed=True)

# distributed training/evaluation/predict
forecaster.fit(...)
forecaster.predict(...)
forecaster.evaluate(...)

# to non distributed version
forecaster.to_local()

# use onnx inference
forecaster.predict_with_onnx(...)
forecaster.evaluate_with_onnx(...)
```

- [x] Distributed fit/evaluate/predict
- [x] Distributed save/load
- [x] ONNX runtime inference after distributed training (`to_local`)
- [x] batchsize parameter are supported for distributed/non-distributed, onnx/non-onnx forecaster in evaluate and predict
- [x] TCNForecaster default hyperparameter's change
- [x] More UT
- [x] `metrics` param

fit(x, y, ...) will be changed to fit(data, ...) later

### CODE REVIEW

File: pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py [hunk 1]
```diff
@@ -205,27 +273,84 @@ def evaluate_with_onnx(self, x, y,
...
 
         :param checkpoint_file: The checkpoint file location you want to load the forecaster.
         """
-        self.internal.restore(checkpoint_file)
+        if self.distributed:
+            self.internal.load(checkpoint_file)
+        else:
+            self.internal.restore(checkpoint_file)
+
+    def to_local(self, cache_dir="/tmp/chronos/tcn_forecaster"):
```
Comment (yushan111) [inline pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py:311]: shall we just use `get_model` in `orca torch estimator`

File: pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py [hunk 2]
```diff
@@ -205,27 +273,84 @@ def evaluate_with_onnx(self, x, y,
...
+        if not self.distributed:
+            raise RuntimeError("A local forecaster does not need shutdown.")
+        self.internal.shutdown(force)
+
+    def _np_to_creator(self, data):
+        def data_creator(config, batch_size):
+                return DataLoader(TensorDataset(torch.from_numpy(data[0]).float(),
+                                                torch.from_numpy(data[1]).float()),
+                                  batch_size=batch_size,
+                                  shuffle=True)
```
Comment (yushan111) [inline pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py:355]: put that to somewhere more common like data_utils?

File: pyzoo/zoo/chronos/model/tcn.py [hunk 3]
```diff
@@ -125,6 +125,7 @@ def init_weights(self):
         self.linear.weight.data.normal_(0, 0.01)
 
     def forward(self, x):
+        x = x.float()
```
Comment (yushan111) [inline pyzoo/zoo/chronos/model/tcn.py:128]: add comment for that

**Review by yushan111 — COMMENTED** [review]
Others LGTM

File: pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py [hunk 4]
```diff
@@ -101,30 +124,64 @@ def __init__(self,
...
-        :param validation_data: A tuple (x_valid, y_valid) as validation data. Default to None.
-        :param epochs: Number of epochs you want to train.
-        :param metric: The metric for training data.
-        :param batch_size: Number of batch size you want to train.
+        :param validation_data: A tuple (x_valid, y_valid) as validation data. Default to None. The
+               value is only effective if the forecaster is in a non-distributed mode. If the
+               forecaster is distributed or validation_data is set to None, forecaster will
+               evaluate on the training data.
+        :param epochs: Number of epochs you want to train. The value defaults to 1.
+        :param metric: The metric for validation on validation_data. The value is only effective
```
Comment (shane-huang) [inline pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py:162]: "validaton_data" and "metric" are only effective in non-distributed mode - it seems not very intuitive to user. I think we should remove these from the Forcaster interface. Later we should refactor BaseModel and seperate fit and fit_eval interfaces.

Comment (shane-huang) [reply pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py:162]: If not necessary, distributed and standalone training should have as little difference as possible.

File: pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py [hunk 5]
```diff
@@ -144,58 +201,115 @@ def _check_data(self, x, y):
...
         :param x: A numpy array with shape (num_samples, lookback, feature_dim).
+        :param batch_size: predict batch size. The value will not affect predict
+               result but will affect resources cost(e.g. memory and time).
 
         :return: A numpy array with shape (num_samples, lookback, feature_dim).
         """
-        if not self.internal.model_built:
-            raise RuntimeError("You must call fit or restore first before calling predict!")
-        return self.internal.predict(x)
+        if self.distributed:
```
Comment (shane-huang) [inline pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py:214]: If training is distributed, is it necessary that prediction is distributed as well? In our previous experience, in many cases training takes a lot of time and needs distributed training while prediction often happens on a single node (single thread or local parallel). Maybe we need an distributed option for predict instead of using self.distributed.

File: pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py [hunk 6]
```diff
@@ -41,14 +52,18 @@ def __init__(self,
...
+                 dropout=0.1,
                  optimizer="Adam",
                  loss="mse",
                  lr=0.001,
-                 seed=None):
+                 seed=None,
+                 distributed=False,
+                 workers_per_node=1,
+                 distributed_backend="torch_distributed",
+                 distributed_matrics=["mse"]):
```
Comment (shane-huang) [inline pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py:66]: shall we set distributed to the class or just to fit? In many cases we only need distributed training while prediciton and evalution happens on small set of data and not need for distributed training.

File: pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py [hunk 7]
```diff
@@ -205,27 +319,73 @@ def evaluate_with_onnx(self, x, y,
...
         """
         Save the forecaster.
 
         :param checkpoint_file: The location you want to save the forecaster.
         """
-        if not self.internal.model_built:
-            raise RuntimeError("You must call fit or restore first before calling save!")
-        self.internal.save(checkpoint_file)
+        if self.distributed:
+            self.internal.save(checkpoint_file)
```
Comment (shane-huang) [inline pyzoo/zoo/chronos/model/forecast/tcn_forecaster.py:341]: what will happen if user don't use fit before calling save in distributed mode?

**Comment by TheaperDeng** [comment]
pr validation combine pass:
http://10.112.231.51:18888/job/ZOO-PR-Validation/6154/
http://10.112.231.51:18888/job/ZOO-PR-Colab-Pytorch/1297/
http://10.112.231.51:18888/job/ZOO-PR-Python-ExampleTests-py37-Spark3/1686/
